File: bindings/typescript/src/assemblers/key-value.ts
import { FragmentId, KeyValueLayoutFragment } from '@vizstack/schema';
import { FragmentAssembler } from '../fragment-assembler';
import _ from 'lodash';

type KeyValueLayoutConfig = {
    separator?: string;
    startMotif?: string;
    endMotif?: string;
    alignSeparators?: boolean;
    showLabels?: boolean;
};

class KeyValueLayoutFragmentAssembler extends FragmentAssembler {
    private _entries: [any, any][] = [];
    private _separator?: string;
    private _startMotif?: string;
    private _endMotif?: string;
    private _alignSeparators?: boolean;
    private _showLabels?: boolean;

    constructor(entries?: [any, any][], config: KeyValueLayoutConfig = {}) {
        super();
        if (entries) this._entries = entries;
        this.config(config);
    }

    public item(key: any, value: any) {
        this._entries.push([key, value]);
        return this;
    }

    public config(config: KeyValueLayoutConfig) {
        const { separator, startMotif, endMotif, alignSeparators, showLabels } = config;
        if (separator !== undefined) this._separator = separator;
        if (startMotif !== undefined) this._startMotif = startMotif;
        if (endMotif !== undefined) this._endMotif = endMotif;
        if (alignSeparators !== undefined) this._alignSeparators = alignSeparators;
        if (showLabels !== undefined) this._showLabels = showLabels;
    }

    public assemble(
        getId: (obj: any, name: string) => FragmentId,
    ): [KeyValueLayoutFragment, any[]] {
        return [
            {
                type: 'KeyValueLayout',
                contents: {
                    entries: this._entries.map(([key, value], idx) => ({
                        key: getId(key, `${idx}k`),
                        value: getId(value, `${idx}v`),
                    })),
                    separator: this._separator,
                    startMotif: this._startMotif,
                    endMotif: this._endMotif,
                    alignSeparators: this._alignSeparators,
                    showLabels: this._showLabels,
                },
                meta: this._meta,
            },
            _.flatMap(this._entries, ([key, value]) => [key, value]),
        ];
    }
}

export function KeyValue(entries?: [any, any][], config: KeyValueLayoutConfig = {}) {
    return new KeyValueLayoutFragmentAssembler(entries, config);
}

export interface KeyValue extends ReturnType<typeof KeyValue> {}

File: bindings/python/vizstack/assemblers/keyvalue.py
from vizstack.fragment_assembler import FragmentAssembler
from typing import Optional, Tuple, Dict, List, Any
from vizstack.schema import Fragment


class KeyValue(FragmentAssembler):

    _separator: Optional[str] = None
    _start_motif: Optional[str] = None
    _end_motif: Optional[str] = None
    _align_separators: Optional[bool] = None
    _show_labels: Optional[bool] = None

    def __init__(self,
                 keyvalues: Optional[Dict[Any, Any]] = None,
                 separator: Optional[str] = None,
                 start_motif: Optional[str] = None,
                 end_motif: Optional[str] = None,
                 align_separators: Optional[bool] = None,
                 show_labels: Optional[bool] = None,
                 ) -> None:
        """"""
        super(KeyValue, self).__init__()
        self._entries: List[Tuple[Any, Any]] = []
        if keyvalues:
            for key, value in keyvalues.items():
                self.item(key, value)
        self.config(separator=separator, start_motif=start_motif, end_motif=end_motif,
                    align_separators=align_separators, show_labels=show_labels)

    def item(self, key: Any, value: Any):
        self._entries.append((key, value))
        return self

    def config(self,
               separator: Optional[str] = None,
               start_motif: Optional[str] = None,
               end_motif: Optional[str] = None,
               align_separators: Optional[bool] = None,
               show_labels: Optional[bool] = None,
               ):
        if separator is not None: self._separator = separator
        if start_motif is not None: self._start_motif = start_motif
        if end_motif is not None: self._end_motif = end_motif
        if align_separators is not None: self._align_separators = align_separators
        if show_labels is not None: self._show_labels = show_labels
        return self

    def assemble(self, get_id) -> Tuple[Fragment, List[Any]]:
        return {
            'type': 'KeyValueLayout',
            'contents': {
                'entries': [
                    {'key': get_id(key, '{}k'.format(i)), 'value': get_id(value, '{}v'.format(i))}
                    for i, (key, value) in enumerate(self._entries)],
                'separator': self._separator,
                'startMotif': self._start_motif,
                'endMotif': self._end_motif,
                'alignSeparators': self._align_separators,
                'showLabels': self._show_labels,

            },
            'meta': self._meta,
        }, [t[0] for t in self._entries] + [t[1] for t in self._entries]
